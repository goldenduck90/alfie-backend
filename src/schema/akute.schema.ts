import { Field, InputType, ObjectType } from "type-graphql"

@InputType()
export class PharmacyLocationInput {
  @Field()
  name: string
}

@ObjectType()
export class PharmacyLocationResult {
  @Field()
  id: number

  @Field()
  name: string

  @Field()
  address_line_1: string

  @Field(() => String, { nullable: true })
  address_line_2: string

  @Field()
  address_city: string

  @Field()
  address_state: string

  @Field()
  address_zipcode: string

  @Field()
  primary_phone_number: string

  @Field()
  primary_fax_number: string

  @Field(() => [String])
  pharmacy_specialties: string[]

  @Field(() => Number, { nullable: true })
  lat: number

  @Field(() => Number, { nullable: true })
  lng: number
}

@ObjectType()
export class CreateLabOrderResponse {
  @Field(() => String)
  labOrderId: string
}

@InputType()
export class DocUploadInput {
  @Field(() => String)
  file: string

  @Field(() => String)
  fileName: string

  @Field(() => String, { nullable: true })
  description?: string

  @Field(() => String, { nullable: true })
  externalPatientId?: string

  @Field(() => String, { nullable: true })
  patientId?: string

  @Field(() => [String])
  tags?: string[]
}

@ObjectType()
export class AkuteDocument {
  @Field(() => String)
  id: string
}
