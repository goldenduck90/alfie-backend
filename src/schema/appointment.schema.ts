import { Field, InputType, ObjectType } from "type-graphql"
import {
  Break,
  Day,
  IEAProvider,
  Settings,
  WorkingPlan,
} from "./../@types/easyAppointmentTypes"
import Role from "./enums/Role"

@ObjectType()
@InputType("TimeBlockInput")
class TimeBlock {
  @Field()
  start: string

  @Field()
  end: string
}

@ObjectType()
@InputType("BreakInput")
class ScheduleBreak extends TimeBlock {}

@ObjectType()
@InputType("DailyScheduleInput")
class DailySchedule extends TimeBlock {
  @Field(() => [ScheduleBreak])
  breaks: ScheduleBreak[]
}

@InputType("ScheduleInput")
@ObjectType()
export class Schedule {
  @Field(() => DailySchedule)
  monday: DailySchedule

  @Field(() => DailySchedule)
  tuesday: DailySchedule

  @Field(() => DailySchedule)
  wednesday: DailySchedule

  @Field(() => DailySchedule)
  thursday: DailySchedule

  @Field(() => DailySchedule)
  friday: DailySchedule

  @Field(() => DailySchedule)
  saturday: DailySchedule

  @Field(() => DailySchedule)
  sunday: DailySchedule
}

@InputType("ScheduleExceptionsInput")
@ObjectType()
class ScheduleExceptions {
  @Field(() => DailySchedule)
  date: DailySchedule
}
@ObjectType()
export class ScheduleObject {
  @Field()
  timezone: string

  @Field(() => Schedule)
  schedule: Schedule

  @Field(() => ScheduleExceptions)
  exceptions: ScheduleExceptions
}

@ObjectType()
export class UpdateScheduleMessage {
  @Field()
  code: number

  @Field()
  message: string
}

@InputType("ScheduleInput2")
@ObjectType()
export class ScheduleObjectInput {
  @Field()
  timezone: string

  @Field(() => Schedule)
  schedule: Schedule
}
@ObjectType()
export class EAService {
  @Field(() => String)
  id: string

  @Field(() => String)
  name: string

  @Field(() => Number)
  durationInMins: number

  @Field(() => String, { nullable: true })
  description?: string
}

@ObjectType()
export class EAProvider {
  @Field(() => String)
  id: string

  @Field(() => String)
  name: string

  @Field(() => String)
  email: string

  @Field(() => Role)
  type: Role

  @Field(() => Number, { nullable: true })
  numberOfPatients?: number

  @Field(() => String)
  timezone: string

  @Field(() => Number)
  minAdvancedNotice: number

  @Field(() => Number)
  bufferTime: number
}

@ObjectType()
export class EACustomer {
  @Field(() => String)
  id: string

  @Field(() => String)
  name: string

  @Field(() => String)
  email: string

  @Field(() => String)
  phone: string
}

@ObjectType()
export class TimeslotsResponse {
  @Field(() => String)
  selectedDate: string

  @Field(() => Number)
  total: number

  @Field(() => String)
  timezone: string

  @Field(() => EAService)
  eaService: EAService

  @Field(() => EAProvider)
  eaProvider: EAProvider

  @Field(() => EACustomer, { nullable: true })
  eaCustomer?: boolean

  @Field(() => [Timeslot])
  timeslots: Timeslot[]
}

@ObjectType()
export class Timeslot {
  @Field(() => String)
  start: string

  @Field(() => String)
  end: string
}

@InputType()
export class GetTimeslotsInput {
  @Field(() => String)
  selectedDate: string

  @Field(() => String)
  timezone: string

  @Field(() => String, { nullable: true })
  appointmentId?: string

  @Field(() => Boolean, { defaultValue: false })
  bypassNotice: boolean

  @Field(() => String, { nullable: true })
  userId?: string
}

@InputType()
export class CreateCustomerInput {
  @Field(() => String)
  userId: string

  @Field(() => String)
  firstName: string

  @Field(() => String)
  lastName: string

  @Field(() => String)
  email: string

  @Field(() => String)
  phone: string

  @Field(() => String)
  address: string

  @Field(() => String)
  city: string

  @Field(() => String)
  state: string

  @Field(() => String)
  zipCode: string

  @Field(() => String)
  timezone: string

  @Field(() => String, { nullable: true })
  notes?: string

  @Field(() => Boolean, { nullable: true })
  updateUser?: boolean
}

@InputType()
export class CreateAppointmentInput {
  @Field(() => String, { nullable: true })
  userId?: string

  @Field(() => String)
  start: string

  @Field(() => String)
  end: string

  @Field(() => String)
  timezone: string

  @Field(() => Boolean, { defaultValue: false })
  bypassNotice: boolean

  @Field(() => String, { nullable: true })
  notes?: string

  @Field(() => String, { nullable: true })
  userTaskId?: string
}

@InputType()
export class UpdateAppointmentInput {
  @Field(() => String)
  eaAppointmentId: string

  @Field(() => String)
  start: string

  @Field(() => String)
  end: string

  @Field(() => String)
  timezone: string

  @Field(() => String, { nullable: true })
  notes?: string

  @Field(() => Boolean, { defaultValue: false })
  bypassNotice: boolean
}

@InputType()
export class GetAppointmentInput {
  @Field(() => String)
  eaAppointmentId: string

  @Field(() => String)
  timezone: string
}

@InputType()
export class UpcomingAppointmentsInput {
  @Field(() => String, { nullable: true })
  selectedDate?: string

  @Field(() => String)
  timezone: string
}

@InputType()
export class GetAppointmentsByMonthInput {
  @Field(() => Number)
  month: number

  @Field(() => String)
  timezone: string
}

@InputType()
export class GetAppointmentsByDateInput {
  @Field(() => String)
  selectedDate: string

  @Field(() => String)
  timezone: string
}

@ObjectType()
export class EAAppointment {
  @Field(() => String)
  eaAppointmentId: string

  @Field(() => EAService)
  eaService: string

  @Field(() => EAProvider)
  eaProvider: EAProvider

  @Field(() => String)
  start: string

  @Field(() => String)
  end: string

  @Field(() => String)
  location: string

  @Field(() => String, { nullable: true })
  notes?: string

  @Field(() => EACustomer)
  eaCustomer: EACustomer

  @Field(() => Boolean)
  attendanceEmailSent: boolean

  @Field(() => Boolean)
  claimSubmitted: boolean

  @Field(() => Boolean)
  patientAttended: boolean

  @Field(() => Boolean)
  providerAttended: boolean
}

@ObjectType()
@InputType("EAProviderProfileWorkingPlanBreakInput")
export class EAWorkingPlanBreak implements Break {
  @Field(() => String, { nullable: true })
  start?: string

  @Field(() => String, { nullable: true })
  end?: string
}

@ObjectType()
@InputType("EAProviderProfileWorkingPlanDayInput")
export class EAWorkingPlanDay implements Day {
  @Field(() => String, { nullable: true })
  start?: string

  @Field(() => String, { nullable: true })
  end?: string

  @Field(() => [EAWorkingPlanBreak], { nullable: true })
  breaks?: EAWorkingPlanBreak[]
}

@ObjectType()
@InputType("EAProviderProfileWorkingPlanInput")
export class EAWorkingPlan implements WorkingPlan {
  @Field(() => EAWorkingPlanDay, { nullable: true })
  monday?: EAWorkingPlanDay

  @Field(() => EAWorkingPlanDay, { nullable: true })
  tuesday?: EAWorkingPlanDay

  @Field(() => EAWorkingPlanDay, { nullable: true })
  wednesday?: EAWorkingPlanDay

  @Field(() => EAWorkingPlanDay, { nullable: true })
  thursday?: EAWorkingPlanDay

  @Field(() => EAWorkingPlanDay, { nullable: true })
  friday?: EAWorkingPlanDay

  @Field(() => EAWorkingPlanDay, { nullable: true })
  saturday?: EAWorkingPlanDay

  @Field(() => EAWorkingPlanDay, { nullable: true })
  sunday?: EAWorkingPlanDay
}

@ObjectType()
@InputType("EAProviderProfileSettingsInput")
export class EAProviderSettings implements Settings {
  @Field(() => EAWorkingPlan, { nullable: true })
  workingPlan?: EAWorkingPlan
}

@ObjectType()
@InputType("EAProviderProfileInput")
export class EAProviderProfile implements IEAProvider {
  @Field(() => String, { nullable: true })
  firstName?: string

  @Field(() => EAProviderSettings, { nullable: true })
  settings?: EAProviderSettings
}
