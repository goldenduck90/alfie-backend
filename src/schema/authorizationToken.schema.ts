import { Field, ObjectType, InputType } from "type-graphql"
import { getModelForClass, prop } from "@typegoose/typegoose"

@ObjectType()
export class AuthorizationToken {
  @Field(() => String)
  @prop({ required: true })
  token: string

  @Field(() => String)
  @prop({ required: false })
  refreshToken?: string

  @Field(() => Date)
  @prop({ required: false })
  expiresAt: Date

  @Field(() => Date, { nullable: true })
  @prop({ required: false })
  refreshTokenExpiresAt?: Date

  @Field(() => String)
  @prop({ required: true })
  provider: string
}

@InputType()
export class RefreshTokenInput {
  @Field(() => String)
  refreshToken: string
}

export const AuthorizationTokenModel = getModelForClass<
  typeof AuthorizationToken
>(AuthorizationToken, {
  schemaOptions: { timestamps: true },
})
