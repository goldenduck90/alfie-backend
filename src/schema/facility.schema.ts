import { ModelOptions, getModelForClass, prop } from "@typegoose/typegoose"
import { Field, ObjectType } from "type-graphql"

@ObjectType()
@ModelOptions({
  schemaOptions: { _id: false, discriminatorKey: "metriportId" },
})
export class Facility {
  @Field(() => String)
  metriportId: string

  @Field(() => [String])
  @prop({ default: [], required: true })
  states: string[]
}

export const FacilityModel = getModelForClass<typeof Facility>(Facility)
