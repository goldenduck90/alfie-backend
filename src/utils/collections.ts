/**
 * Creates a map for a collection using a function to get the key
 * field by which to reference each element. The result is a map with
 * `{ [retrievedKey: string]: T }`. Useful for collections where the key
 * field is unique to each element.
 */
export const mapCollectionByField = <T>(
  collection: T[],
  getField: (item: T) => string
): Record<string, T> =>
  collection.reduce((map, item) => ({ ...map, [getField(item)]: item }), {})

/** Maps a collection by grouping items by a calculated field from `getField`. */
export const groupCollectionByField = <T>(
  collection: T[],
  getField: (item: T) => string
): Record<string, T[]> =>
  collection.reduce((map, item) => {
    const field = getField(item)
    return { ...map, [field]: [...(map[field] ?? []), item] }
  }, {} as Record<string, T[]>)

/**
 * Remove duplicate entries from items, using a function to extract a comparable string from each item.
 * Prioritizes entries earlier in the list.
 */
export const uniqueItems = <T>(
  items: T[],
  getComparisonValue: (item: T) => string
) => {
  const map: Record<string, boolean> = {}
  const results: T[] = []
  items.forEach((item) => {
    const value = getComparisonValue(item)
    if (!map[value]) {
      results.push(item)
      map[value] = true
    }
  })

  return results
}

/**
 * Sorts the items by an extracted value.
 */
export const sorted = <T>(
  items: T[],
  /** Extract a property that is comparable from an element in `items`. */
  getComparisonValue: (item: T) => any,
  order: "ascending" | "descending" = "ascending"
) =>
  items
    // add the computed comparison value
    .map((item) => ({ item, value: getComparisonValue(item) }))
    // sort by the value
    .sort(
      ({ value: a }, { value: b }) =>
        (b > a ? 1 : b === a ? 0 : -1) * (order === "ascending" ? -1 : 1)
    )
    // map to the item, without the computed value
    .map(({ item }) => item)

/** Returns an array with the `subArray` repeated `times` times. */
export const repeat = <T>(subArray: T[], times: number) => {
  if (times < 0)
    throw new TypeError("repeat `times` argument must be non-negative.")

  return new Array(times)
    .fill(null)
    .map(() => [...subArray])
    .flatMap((x) => x)
}
