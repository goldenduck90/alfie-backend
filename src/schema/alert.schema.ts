import { Ref, getModelForClass, prop } from "@typegoose/typegoose"
import { Field, ObjectType, registerEnumType } from "type-graphql"
import { Task } from "./task.schema"
import { User } from "./user.schema"

export enum SeverityType {
  LOW = "LOW",
  MEDIUM = "MEDIUM",
  HIGH = "HIGH",
  EXTREME = "EXTREME",
}

registerEnumType(SeverityType, {
  name: "SeverityType",
  description: "Alert severity type.",
})

@ObjectType()
export class Alert {
  @Field(() => String)
  _id: string

  @Field(() => String)
  @prop({ required: true })
  title: string

  @Field(() => String)
  @prop({ required: true })
  description: string

  @Field(() => Task)
  @prop({ ref: () => Task, required: true })
  task: Ref<Task>

  @Field(() => User)
  @prop({ ref: () => User, required: true })
  user: Ref<User>

  @Field(() => SeverityType)
  @prop({ enum: SeverityType, required: true })
  severity: SeverityType

  @Field(() => Boolean)
  @prop({ required: true })
  medical: boolean

  @Field(() => Date, { nullable: true })
  @prop({ required: false })
  acknowledgedAt?: Date

  @Field(() => Date, { nullable: true })
  @prop({ required: false })
  notifiedAt?: Date
}

export const AlertModel = getModelForClass<typeof Alert>(Alert, {
  schemaOptions: { timestamps: true },
})
