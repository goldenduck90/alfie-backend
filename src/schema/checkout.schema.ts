import {
  getModelForClass,
  prop,
  ReturnModelType,
  queryMethod,
  index,
} from "@typegoose/typegoose"
import { AsQueryMethod, Ref } from "@typegoose/typegoose/lib/types"
import { IsEmail, IsPhoneNumber } from "class-validator"
import { Field, InputType, ObjectType } from "type-graphql"
import config from "config"
import { Address, Gender, User } from "./user.schema"

const { email: emailValidation, phone: phoneValidation } =
  config.get("validations")

function findByEmail(
  this: ReturnModelType<typeof Checkout, QueryHelpers>,
  email: Checkout["email"]
) {
  return this.findOne({ email })
}

function findByStripePaymentLinkId(
  this: ReturnModelType<typeof Checkout, QueryHelpers>,
  stripePaymentLinkId: Checkout["stripePaymentLinkId"]
) {
  return this.findOne({ stripePaymentLinkId })
}

interface QueryHelpers {
  findByEmail: AsQueryMethod<typeof findByEmail>
  findByStripePaymentLinkId: AsQueryMethod<typeof findByStripePaymentLinkId>
}

@index({ email: 1 })
@index({ stripePaymentLinkId: 1 })
@queryMethod(findByEmail)
@queryMethod(findByStripePaymentLinkId)
@ObjectType()
export class Checkout {
  @Field(() => String)
  _id: string

  @Field(() => String)
  @prop({ required: true })
  stripePaymentLinkId: string

  @Field(() => String)
  @prop({ required: true })
  name: string

  @Field(() => String)
  @prop({ required: true })
  email: string

  @Field(() => String)
  @prop({ required: true })
  weightLossMotivator: string

  @Field(() => Date)
  @prop({ required: true })
  dateOfBirth: Date

  @Field(() => Gender)
  @prop({ enum: Gender, type: String, required: true })
  gender: Gender

  @Field(() => String)
  @prop({ required: true })
  state: string

  @Field(() => Number)
  @prop({ required: true })
  heightInInches: number

  @Field(() => Number)
  @prop({ required: true })
  weightInLbs: number

  @Field(() => Boolean)
  @prop({ default: false, required: true })
  checkedOut: boolean

  @Field(() => String)
  @prop()
  stripeCheckoutId?: string

  @Field(() => User, { nullable: true })
  @prop({ ref: "User" })
  user?: Ref<User>
}

export const CheckoutModel = getModelForClass<typeof Checkout, QueryHelpers>(
  Checkout,
  {
    schemaOptions: { timestamps: true },
  }
)

@InputType()
export class CreateCheckoutInput {
  @Field(() => String)
  name: string

  @IsEmail({}, { message: emailValidation.message })
  @Field(() => String)
  email: string

  @Field(() => String)
  weightLossMotivator: string

  @Field(() => Date)
  dateOfBirth: Date

  @Field(() => Gender)
  gender: Gender

  @Field(() => String)
  state: string

  @Field(() => Number)
  heightInInches: number

  @Field(() => Number)
  weightInLbs: number
}

@InputType()
export class GetCheckoutInput {
  @Field(() => String)
  _id: string
}
@InputType()
export class CompleteCheckoutInput {
  @Field(() => String, { description: "Stripe payment link ID" })
  stripePaymentLinkId: string

  @Field(() => String, { description: "Stripe subscription ID" })
  stripeSubscriptionId: string

  @Field(() => String, { description: "Stripe customer ID" })
  stripeCustomerId: string

  @Field(() => String, { description: "Stripe checkout ID" })
  stripeCheckoutId: string

  @IsPhoneNumber("US", { message: phoneValidation.message })
  @Field(() => String)
  phone: string

  @Field(() => Address, { description: "Retrieved from Stripe checkout" })
  address: Address

  @Field(() => Date, {
    nullable: true,
    description:
      "When the user's subscription expires. Retrieved from Stripe checkout. Default is 1 month from now.",
  })
  subscriptionExpiresAt?: Date
}

@ObjectType()
export class CheckoutResponse {
  @Field(() => String, { nullable: true })
  message: string

  @Field(() => Checkout)
  checkout: Checkout

  @Field(() => String)
  paymentLink: string
}
