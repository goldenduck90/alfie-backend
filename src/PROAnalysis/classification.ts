import {
  Classification,
  ClassificationType,
  Score,
} from "../schema/user.schema"
import { TaskType } from "../schema/task.schema"
import { calculateSetting, calculateSettings } from "../utils/calculateSetting"

/**
 * Given this array of objects we need to classify a patient as each of the following and their percentile for each classification:
 * 1. Growler
 * 2. Empath
 * 3. Ember
 * 4. Rover

 * The criteria for each classification is as follows:

 * Empath:
 * If the HADS(task: MP_FEELING) Anxiety Score is in the top 75th percentile, the person is classified as an Empath

 * Ember:
 * If the person’s activity (task: MP_ACTIVITY) level is in the bottom 25th percentile, they are classified as an ember
 * If bottom 25th percentile then really we need to display 75th percentile

 * Growler:
 * If the person is in the bottom 25th percentile for satiety(hunger at 2 hours) (task: MP_HUNGER), they are a growler (i.e.they don’t feel full)
 * If bottom 25th percentile then really we need to display 75th percentile

 * Rover:
 * If the person is in the top 25 % for ad libitum (task: AD_LIBITUM) calories eaten, they are classified as a rover
 * If the person is in the bottom 25 % for hunger(30 minutes) (task: MP_HUNGER), they can also be classified as rover
 * If your bottom 25th percentile then really we need to display 75th percentile
 */
export function classifyUser(scores: Score[]): Classification[] {
  const classifications: Classification[] = []

  for (const score of scores) {
    switch (score.task) {
      case TaskType.MP_ACTIVITY:
      case TaskType.AD_LIBITUM:
      case TaskType.MP_FEELING: {
        const { classificationType, inverse } =
          calculateSetting<{
            classificationType: ClassificationType
            inverse: boolean
          }>(
            classificationCriteria,
            ["classificationType", "inverse"],
            score
          )

        if (classificationType) {
          classifications.push({
            date: score.date,
            classification: classificationType,
            percentile: inverse
              ? inversePercentile(score.percentile)
              : score.percentile,
            calculatedPercentile: inverse
              ? inversePercentile(score.calculatedPercentile)
              : score.calculatedPercentile
          })
        }
        break
      }
      case TaskType.MP_HUNGER: {
        const { classificationType, inverse } =
          calculateSettings<{
            classificationType: ClassificationType[]
            inverse: boolean[]
          }>(
            classificationCriteria,
            ["classificationType", "inverse"],
            score
          )

        const hungerClassifications: Classification[] =
          classificationType.map((classification, index) => ({
            classification,
            date: score.date,
            percentile: inverse[index] ? inversePercentile(score.percentile) : score.percentile,
            calculatedPercentile: inverse[index] ? inversePercentile(score.calculatedPercentile) : score.calculatedPercentile,
          }))

        classifications.push(...hungerClassifications)
        break
      }
      default:
        break
    }
  }

  return classifications
}

/**
 * Returns the inverse of a percentile. For example, the inverse of 80 is 20, or `100 - p`.
 * Returns the percentile in the same type passed, string or number.
 */
function inversePercentile(percentile: string): string;
function inversePercentile(percentile: number): number;
function inversePercentile(percentile: string | number): string | number {
  if (typeof percentile === "string") {
    return `${Math.min(99, 100 - parseInt(percentile))}`
  } else if (typeof percentile === "number") {
    return Math.min(100 - percentile, 99)
  } else {
    return percentile
  }
}

export const classificationCriteria = [
  {
    vars: { classificationType: ClassificationType.Empath, inverse: false },
    conditions: [{
      task: TaskType.MP_FEELING,
      calculatedPercentile: { range: [75, 100] },
    }],
  },
  {
    vars: { classificationType: ClassificationType.Ember, inverse: true },
    conditions: [{
      task: TaskType.MP_ACTIVITY,
      calculatedPercentile: { range: [0, 25] },
    }],
  },
  {
    vars: { classificationType: ClassificationType.Growler, inverse: true },
    conditions: [{
      task: TaskType.MP_HUNGER,
      calculatedPercentile1Hour: { range: [0, 25] },
    }],
  },
  {
    vars: { classificationType: ClassificationType.Rover, inverse: true },
    conditions: [
      { task: TaskType.MP_HUNGER, calculatedPercentile30Minutes: { range: [0, 25] } },
    ],
  },
  {
    vars: { test: 10, classificationType: ClassificationType.Rover, inverse: false },
    conditions: [
      { task: TaskType.AD_LIBITUM, calculatedPercentile: { range: [75, 100] } },
    ],
  }
]
