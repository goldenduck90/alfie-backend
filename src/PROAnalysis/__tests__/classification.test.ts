import { TaskType } from "../../schema/task.schema"
import { ClassificationType, Score } from "../../schema/user.schema"
import { sorted } from "../../utils/collections"
import { classifyUser } from "../classification"

// Notes:
// - If someone is classified as multiple, then list all of the ones they are
//   classified as, as well as the percentile (I believe we gave the mean and
//   SD, so you can calculate the percentile from these assuming it’s normal)

describe("classification", () => {
  const dates = ["01", "03", "05", "07", "09", "11", "13", "15", "17"]
    .map((day) => new Date(`2023-07-${day}T00:00:00.000Z`))

  /**
   * Rover: 
   * If the person is in the top 25% for ad libitum calories eaten, they are classified as a rover
   */
  describe("Should classify Rover from Ad Libitum correctly", () => {

    const adLibitumTests: { score: Score; expected: ClassificationType }[] = ([
      [dates[0], 25, 25, null],
      [dates[1], 15, 25, null],
      [dates[2], 50, 50, null],
      [dates[3], 80, 75, ClassificationType.Rover],
      [dates[3], 74, 75, null],
      [dates[4], 60, 50, null],
      [dates[5], 90, 99, ClassificationType.Rover],
    ] as [Date, number, number, ClassificationType | null][])
      .map(([date, calculatedPercentile, percentile, expected]) => ({
        score: {
          task: TaskType.AD_LIBITUM,
          date,
          calculatedPercentile,
          percentile,
        } as Score,
        expected,
      }))

    it.each(adLibitumTests)("Classification ($date, $percentile, $calculatedPercentile)", ({ score, expected }) => {
      const classifications = classifyUser([score])
      expect(classifications).toMatchSnapshot()
      if (expected) {
        expect(classifications).toHaveLength(1)
        expect(classifications[0].classification).toEqual(expected)
      } else {
        expect(classifications).toHaveLength(0)
      }
    })
  })

  /**
   * Growler:
   * - If the person is in the bottom 25th percentile for satiety
   *   (hunger at 2 hours), they are a growler (i.e. they don't feel full)
   * Rover:
   * - If the person is in the bottom 25% for hunger (30 minutes),
   *   they can also be classified as rover
   */
  describe("Should classify Growler/Rover from Hunger correctly", () => {
    const dataPoints = ([
      [dates[0], 0, 0, 5, 8, [ClassificationType.Growler, ClassificationType.Rover]],
      [dates[1], 25, 0, 30, 7, [ClassificationType.Growler]],
      [dates[2], 0, 25, 15, 26, [ClassificationType.Rover]],
      [dates[3], 50, 50, 60, 45, []],
      [dates[4], 75, 50, 80, 63, []],
      [dates[5], 50, 75, 48, 82, []],
      [dates[6], 75, 75, 80, 67, []],
    ] as [Date, number, number, number, number, ClassificationType[]][])

    const hungerTests: { score: Score; expected: ClassificationType[] }[] = dataPoints
      .map(([date, percentile30Minutes, percentile1Hour, calculatedPercentile30Minutes, calculatedPercentile1Hour, expected]) => ({
        score: {
          task: TaskType.MP_HUNGER,
          date,
          calculatedPercentile30Minutes,
          calculatedPercentile1Hour,
          calculatedPercentile: calculatedPercentile30Minutes,
          percentile1Hour,
          percentile30Minutes,
        } as Score,
        expected,
      }))

    it.each(hungerTests)("Classification ($score.date, $score.calculatedPercentile30Minutes, $score.calculatedPercentile1Hour)", ({ score, expected }) => {
      const classifications = classifyUser([score])
      expect(classifications).toMatchSnapshot()
      const classificationTypes = classifications
        .map(({ classification }) => classification)
      expect(sorted(classificationTypes, (x) => x, "ascending"))
        .toEqual(expected)
    })
  })

  /**
   * Ember: 
   * If the person’s activity level is in the bottom 25th percentile,
   * they are classified as an ember.
   */
  describe("Should classify Ember from Activity correctly", () => {
    const activityTests: { score: Score; expected: ClassificationType | null }[] = ([
      [dates[0], 75, 80, null],
      [dates[1], 25, 30, null],
      [dates[2], 50, 60, null],
      [dates[3], 0, 5, ClassificationType.Ember],
      [dates[3], 25, 24, ClassificationType.Ember],
    ] as [Date, number, number, ClassificationType | null][])
      .map(([date, percentile, calculatedPercentile, expected]) => ({
        score: {
          date,
          task: TaskType.MP_ACTIVITY,
          percentile,
          calculatedPercentile,
        },
        expected,
      }))

      it.each(activityTests)("Classification ($score.date, $score.percentile, $score.calculatedPercentile)", ({ score, expected }) => {
        const classifications = classifyUser([score])
        expect(classifications).toMatchSnapshot()
        if (expected) {
          expect(classifications).toHaveLength(1)
          expect(classifications[0].classification).toEqual(expected)
        } else {
          expect(classifications).toHaveLength(0)
        }
      })
  })

  /**
   * Empath:
   * If the HADS Anxiety Score is in the top 75th percentile,
   * the person is classified as an Empath
   */
  describe("Should classify Empath from Feeling Task", () => {
    const feelingTests = ([
      [dates[0], 0, 5, null],
      [dates[1], 25, 30, null],
      [dates[2], 50, 60, null],
      [dates[3], 50, 76, ClassificationType.Empath],
      [dates[3], 75, 74, null],
      [dates[3], 75, 76, ClassificationType.Empath],
    ] as [Date, number, number, ClassificationType | null][])
      .map(([date, percentile, calculatedPercentile, expected]) => ({
        score: {
          date,
          percentile,
          calculatedPercentile,
          task: TaskType.MP_FEELING,
        } as Score,
        expected
      }))
    
    it.each(feelingTests)("Classification ($score.date, $score.percentile, $score.calculatedPercentile)", ({ score, expected }) => {
      const classifications =  classifyUser([score])
      if (!expected) {
        expect(classifications).toHaveLength(0)
      } else {
        expect(classifications).toHaveLength(1)
        expect(classifications[0].classification).toEqual(expected)
      }
    })
  })
})
