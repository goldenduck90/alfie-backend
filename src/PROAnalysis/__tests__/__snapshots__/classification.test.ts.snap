// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`classification Should classify Ember from Activity correctly Classification (2023-07-01T00:00:00.000Z, 75, 80) 1`] = `[]`;

exports[`classification Should classify Ember from Activity correctly Classification (2023-07-03T00:00:00.000Z, 25, 30) 1`] = `[]`;

exports[`classification Should classify Ember from Activity correctly Classification (2023-07-05T00:00:00.000Z, 50, 60) 1`] = `[]`;

exports[`classification Should classify Ember from Activity correctly Classification (2023-07-07T00:00:00.000Z, 0, 5) 1`] = `
[
  {
    "calculatedPercentile": 95,
    "classification": "Ember",
    "date": 2023-07-07T00:00:00.000Z,
    "percentile": 99,
  },
]
`;

exports[`classification Should classify Ember from Activity correctly Classification (2023-07-07T00:00:00.000Z, 25, 24) 1`] = `
[
  {
    "calculatedPercentile": 76,
    "classification": "Ember",
    "date": 2023-07-07T00:00:00.000Z,
    "percentile": 75,
  },
]
`;

exports[`classification Should classify Growler/Rover from Hunger correctly Classification (2023-07-01T00:00:00.000Z, 5, 8) 1`] = `
[
  {
    "calculatedPercentile": 95,
    "classification": "Growler",
    "date": 2023-07-01T00:00:00.000Z,
    "percentile": undefined,
  },
  {
    "calculatedPercentile": 95,
    "classification": "Rover",
    "date": 2023-07-01T00:00:00.000Z,
    "percentile": undefined,
  },
]
`;

exports[`classification Should classify Growler/Rover from Hunger correctly Classification (2023-07-03T00:00:00.000Z, 30, 7) 1`] = `
[
  {
    "calculatedPercentile": 70,
    "classification": "Growler",
    "date": 2023-07-03T00:00:00.000Z,
    "percentile": undefined,
  },
]
`;

exports[`classification Should classify Growler/Rover from Hunger correctly Classification (2023-07-05T00:00:00.000Z, 15, 26) 1`] = `
[
  {
    "calculatedPercentile": 85,
    "classification": "Rover",
    "date": 2023-07-05T00:00:00.000Z,
    "percentile": undefined,
  },
]
`;

exports[`classification Should classify Growler/Rover from Hunger correctly Classification (2023-07-07T00:00:00.000Z, 60, 45) 1`] = `[]`;

exports[`classification Should classify Growler/Rover from Hunger correctly Classification (2023-07-09T00:00:00.000Z, 80, 63) 1`] = `[]`;

exports[`classification Should classify Growler/Rover from Hunger correctly Classification (2023-07-11T00:00:00.000Z, 48, 82) 1`] = `[]`;

exports[`classification Should classify Growler/Rover from Hunger correctly Classification (2023-07-13T00:00:00.000Z, 80, 67) 1`] = `[]`;

exports[`classification Should classify Rover from Ad Libitum correctly Classification ($date, $percentile, $calculatedPercentile) 1`] = `[]`;

exports[`classification Should classify Rover from Ad Libitum correctly Classification ($date, $percentile, $calculatedPercentile) 2`] = `[]`;

exports[`classification Should classify Rover from Ad Libitum correctly Classification ($date, $percentile, $calculatedPercentile) 3`] = `[]`;

exports[`classification Should classify Rover from Ad Libitum correctly Classification ($date, $percentile, $calculatedPercentile) 4`] = `
[
  {
    "calculatedPercentile": 80,
    "classification": "Rover",
    "date": 2023-07-07T00:00:00.000Z,
    "percentile": 75,
  },
]
`;

exports[`classification Should classify Rover from Ad Libitum correctly Classification ($date, $percentile, $calculatedPercentile) 5`] = `[]`;

exports[`classification Should classify Rover from Ad Libitum correctly Classification ($date, $percentile, $calculatedPercentile) 6`] = `[]`;

exports[`classification Should classify Rover from Ad Libitum correctly Classification ($date, $percentile, $calculatedPercentile) 7`] = `
[
  {
    "calculatedPercentile": 90,
    "classification": "Rover",
    "date": 2023-07-11T00:00:00.000Z,
    "percentile": 99,
  },
]
`;
